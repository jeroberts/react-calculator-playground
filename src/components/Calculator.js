import React, {Component} from 'react';

import ScreenComponent from './ScreenComponent';
import KeypadComponent from './KeypadComponent';

class Calculator extends Component {

  bill = 'bill';

  state = {
    result: ""
  };

  onClickHandler = (buttonClicked) => {
    if (buttonClicked === 'C') {
      this._handleClear();
    } else if (buttonClicked === '=') {
      this._handleCalculate();
    } else if (buttonClicked === '.' ) {
      // Ensure that decimal numbers stay valid with only one decimal point
      const splitResult = this.state.result.split(/\+|-|x|\//);
      if (splitResult[splitResult.length - 1].indexOf('.') === -1) {
        this._handleInput(buttonClicked);
      }
    } else {
      this._handleInput(buttonClicked);
    }
  }

  _handleInput = (input) => {
    this.setState({
      result: this.state.result + input
    });
  }

  _handleClear = () => {
    this.setState({
      result: ""
    });
  }

  _handleCalculate = () => {
    const lastCharacterOfExpression = this.state.result.slice(-1);
    const expressionEndsWithNumber = !isNaN(lastCharacterOfExpression);
    const experssionEndsWithPeriod = lastCharacterOfExpression === '.';

    if (expressionEndsWithNumber || experssionEndsWithPeriod) {
      // Change all 'x' into '*' so that expression is valid
      const validExpression = this.state.result.replace(/x/g, '*');

      this.setState({
        result: eval(validExpression).toString()
      });
    }
  }

  render() {
    return (
      <div className="calculator">
        <ScreenComponent result={this.state.result}/>
        <KeypadComponent onClickHandler={this.onClickHandler}/>
      </div>
    );
  }
}

export default Calculator;
