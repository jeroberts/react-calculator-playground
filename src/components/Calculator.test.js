import React from 'react';
import { shallow } from 'enzyme';
import Calculator from './Calculator';

describe('Testing Calculator component', () => {

    const wrapper = shallow(<Calculator />)

    test('Test that Calculator component renders', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('Test that Calculator._handleInput adds input character to result', () => {
        wrapper.setState({ result: '123' })
        wrapper.instance()._handleInput('+');
    
        expect(wrapper.state('result')).toEqual('123+');
    });

    test('Test that Calculator._handleClear clears result', () => {
        wrapper.setState({result: '12+34'});
        wrapper.instance()._handleClear();

        expect(wrapper.state('result')).toEqual('');
    });

    test('Test that Calculator._handleCalculate calculates result ending in number', () => {
        wrapper.setState({result: '12+34'});
        wrapper.instance()._handleCalculate();

        expect(wrapper.state('result')).toEqual('46');
    });

    test('Test that Calculator._handleCalculate calculates result ending with \'.\'', () => {
        wrapper.setState({result: '12.4+34.'});
        wrapper.instance()._handleCalculate();

        expect(wrapper.state('result')).toEqual('46.4');
    });

    test('Test that Calculator._handleCalculate does nothing given result ends with operand', () => {
        wrapper.setState({result: '12x'});
        wrapper.instance()._handleCalculate();

        expect(wrapper.state('result')).toEqual('12x');
    });

    test('Test that Calculator.onClickHandler clears result given input is \'C\'', () => {
        wrapper.setState({result: '123x234'});
        wrapper.instance().onClickHandler('C');

        expect(wrapper.state('result')).toEqual('');
    });

    test('Test that Calculator.onClickHandler calculates result given input is \'=\'', () => {
        wrapper.setState({result: '3x2'});
        wrapper.instance().onClickHandler('=');

        expect(wrapper.state('result')).toEqual('6');
    });

    test('Test that Calculator.onClickHandler adds \'.\' to result given input is \'.\' and doesn\'t already exist in expresion', () => {
        wrapper.setState({result: '1x2'});
        wrapper.instance().onClickHandler('.');

        expect(wrapper.state('result')).toEqual('1x2.');
    });

    test('Test that Calculator.onClickHandler does not add \'.\' to result given input is \'.\' and expresion already contains a \'.\'', () => {
        wrapper.setState({result: '1x2.'});
        wrapper.instance().onClickHandler('.');

        expect(wrapper.state('result')).toEqual('1x2.');
    });

    test('Test that Calculator.onClickHandler adds the input to result', () => {
        wrapper.setState({result: '1x2'});
        wrapper.instance().onClickHandler('4');

        expect(wrapper.state('result')).toEqual('1x24');
    });
});

