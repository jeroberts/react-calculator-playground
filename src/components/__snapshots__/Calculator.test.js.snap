// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Calculator component Test that Calculator component renders 1`] = `ShallowWrapper {}`;
